src/User.jsx

```diff
@@ -0,0 +1,263 @@
+    if(key) {
+      iframe.postMessage(JSON.stringify({action: 'getKey', key: key}), walletHost);
+    }
+    else {
+      iframe.postMessage(JSON.stringify({action: 'getAllKeys'}), walletHost);
+    }
+
+    var f = function(event){
+
+      if (`${event.origin}/weba-wallet` !== walletHost)
+      return;
+  
+      if(event.data.privateKey) {
+        const address = getAddressFromMnemonic(event.data.privateKey);
+        if(address) {
+          setAddress(address);
+        }
+        window.removeEventListener("message", f, false);
+        iframe.close();
+      }
+      else {
+        window.removeEventListener("message", f, false);
+        iframe.close();
+      }
+    }
+    window.addEventListener("message", f);
+  }
+
+  // function for sending data to wallet
+  const sendDataToWallet = async (key, value) => {
+
+    walletMessenger = window.open(walletHost, 'walletMessenger')
+
+    var f = function(event){
+      if (`${event.origin}/weba-wallet` !== walletHost)
+        return;
+
+      if(event.data == 'received') {
+        sendData(key, value);
+        window.removeEventListener("message", f, false);
+      }
+    }
+    window.addEventListener("message", f);
+  }
+
+  const sendData = async (key, value) => {
+    var message = JSON.stringify({'action': 'storeKey', 'key': key, 'value': value});
+    walletMessenger.postMessage(message, walletHost);
+
+    var f = function(event){
+      if (`${event.origin}/weba-wallet` !== walletHost)
+        return;
+
+      window.removeEventListener("message", f, false);
+      walletMessenger.close();
+      }
+      window.addEventListener("message", f);
+  }
+
+  const metaMaskLogin = async e => {
+    e.preventDefault();
+    e.stopPropagation(); 
+    if (address) {
+      toggleOpen('user');
+    } else {
+      if (!loggingIn) {
+        setLoggingIn(true);
+        try {
+
+          const {address, profile} = await ceramicApi.login();
+          setAddress(address);
+          setShow(false);
+
+        } catch (err) {
+          console.warn(err);
+        } finally {
+          setLoggingIn(false);
+        }
+      }
+    }
+  }
+
+  const discordLogin = async e => {
+    e.preventDefault();
+    e.stopPropagation();
+    setLoginButtons(false);
+  } 
+  
+  useEffect(async () => {
+
+      const {
+        error,
+        error_description,
+        code,
+        id,
+        play,
+        realmId,
+        twitter: arrivingFromTwitter,
+      } = typeof window !== 'undefined' ? parseQuery(window.location.search) : {};
+      if(code) {
+        const {address, error, mnemonic} = await handleDiscordLogin(code, id);
+        if(address) {
+          setAddress(address);
+          sendDataToWallet('privateKey', mnemonic)
+          setShow(false);
+        }
+        else {
+          setLoginError(String(error).toLocaleUpperCase());
+        }
+      }
+      else {
+        fetchWalletData('privateKey');
+      }
+  }, [address, setAddress]);
```

Comment:
I think this `setAddress` is probably pointless?

Reply:
Actually never mind, I can see how the way the code is structured it could be needed!